### PR TITLE
fix(components): the label causes the click event to trigger twice

### LINKED ISSUE #9981: [Component] [checkbox] Checkbox is not working inside label tag
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.40`
- Element Plus Version: `2.2.17`
- Browser / OS: `Chrome 105, Firefox 105 / Windows 10`
- Build Tool: `Vue CLI`

### Related Component

- `el-checkbox`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IEVsZW1lbnRQbHVzIH0gZnJvbSAnQGVsZW1lbnQtcGx1cy9pY29ucy12dWUnXG5cbiAgY29uc3QgdmFsdWUgPSByZWYodHJ1ZSlcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxlbC1mb3JtPlxuICBcdDxlbC1mb3JtLWl0ZW0+XG4gIFx0XHQ8bGFiZWw+XG4gICAgXHRcdDxlbC1jaGVja2JveCB2LW1vZGVsPVwidmFsdWVcIi8+XG4gICAgXHQ8L2xhYmVsPlxuICBcdDwvZWwtZm9ybS1pdGVtPlxuICA8L2VsLWZvcm0+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

When el-checkbox with no label prop is wrapped with label tag it's not possible to change it's state

### What is Expected?

checkbox change it's state on click

### What is actually happening?

checkbox ignores clicks

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by iDestin**
Please make sure these boxes are checked before submitting your PR, thank you!

fixed #9981 

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by chenxch** [comment]
Need to solve the problem of unit testing first.

**Comment by iDestin** [comment]
> Need to solve the problem of unit testing first.

done, The previous method is not very good. I changed it

**Review by chenxch — APPROVED** [review]
Good job.

**Comment by chenxch** [comment]
I seem to have found a scene,e.g.
```vue
<label>
  <el-checkbox v-model="value"><label>text</label></el-checkbox>
</label>
```
When you click `text`, the event will not be triggered. Of course, this has nothing to do with this repair.

**Comment by iDestin** [comment]
You're right, but I found that the current version cannot be triggered

[demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IEVsZW1lbnRQbHVzIH0gZnJvbSAnQGVsZW1lbnQtcGx1cy9pY29ucy12dWUnXG5cbmNvbnN0IHZhbHVlID0gcmVmKHRydWUpXG5jb25zdCB2YWx1ZTEgPSByZWYodHJ1ZSlcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxlbC1mb3JtPlxuICBcdDxlbC1mb3JtLWl0ZW0+XG4gIFx0XHQ8bGFiZWw+XG4gICAgXHRcdDxlbC1jaGVja2JveCB2LW1vZGVsPVwidmFsdWVcIi8+XG4gICAgXHQ8L2xhYmVsPlxuICBcdDwvZWwtZm9ybS1pdGVtPlxuICA8L2VsLWZvcm0+XG4gIFxuICA8bGFiZWw+XG4gICAgPGVsLWNoZWNrYm94IHYtbW9kZWw9XCJ2YWx1ZTFcIj48bGFiZWw+dGV4dDwvbGFiZWw+PC9lbC1jaGVja2JveD5cbiAgPC9sYWJlbD5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

**Comment by chenxch** [comment]
Well, this is a leftover problem.

**Comment by makedopamine** [comment]
I wasn't quite follow you. I mean, we don't  need to check whether a checkbox is wrapped with label first.

**Comment by iDestin** [comment]
I'm not sure if there is a better way. Maybe you can propose a pr @makedopamine

**Comment by makedopamine** [comment]
well, I will try it later.

**Comment by holazz** [comment]
I think a better way is to stop propagation in the inner element's click event when the event's path contains `label`, something like:

```vue
<span :class="ns.e('inner')" @click="handleInnerClick" />
```

```ts
const handleInnerClick = (e: MouseEvent) => {
  const eventTargets: EventTarget[] = e.composedPath()
  if (eventTargets.some((item) => (item as HTMLElement).tagName === 'LABEL')) {
    e.stopPropagation()
  }
}
```

WDYT? @iDestin @makedopamine @chenxch

**Comment by chenxch** [comment]
> WDYT? @iDestin @makedopamine @chenxch

Good idea, part of the idea in this #10003  is right.

**Comment by chenxch** [comment]
@holazz This needs to remove the click event on the root.

**Comment by holazz** [comment]
> @holazz This needs to remove the click event on the root.

Why does it need to be removed?🤔 Is there something I haven't considered?

**Comment by chenxch** [comment]
> Why does it need to be removed?🤔 Is there something I haven't considered?

I modified the code according to your scheme and removed the click event on the root with reference to #10003 and everything seems to be working fine.

**Comment by chenxch** [comment]
The change event will fire twice.

**Comment by chenxch** [comment]
image
Seems like that's all it takes. .

**Comment by iDestin** [comment]
> WDYT? @iDestin @makedopamine @chenxch

cool 👍🏻

**Comment by holazz** [comment]
> The change event will fire twice.

You are right, I missed it. I think the function `onClickRoot` should not be removed directly. It seems that the judgment condition should still be placed in function `onClickRoot`, like this PR currently.

**Comment by holazz** [comment]
So this PR doesn't seem to need any more changes at this time, WDYT? @chenxch

**Comment by chenxch** [comment]
> So this PR doesn't seem to need any more changes at this time, WDYT? @chenxch

Yes, at least I think so now.

File: packages/components/checkbox/src/composables/use-checkbox-event.ts [hunk 1]
```diff
@@ -52,11 +52,19 @@ export const useCheckboxEvent = (
     if (isLimitExceeded.value) return
 
     if (!hasOwnLabel.value && !isDisabled.value && isLabeledByFormItem.value) {
-      model.value = getLabeledValue(
-        [false, props.falseLabel].includes(model.value)
+      // fix: https://github.com/element-plus/element-plus/issues/9981
+      const eventTargets: Array<EventTarget> =
+        e.composedPath && e.composedPath()
```
Comment (holazz) [inline packages/components/checkbox/src/composables/use-checkbox-event.ts:57]: Is the judgment here necessary? How about simplifying it like this:
```suggestion
        e.composedPath()
```

File: packages/components/checkbox/src/composables/use-checkbox-event.ts [hunk 2]
```diff
@@ -52,11 +52,19 @@ export const useCheckboxEvent = (
...
+      const hasLabel = eventTargets.some(
+        (item) => (item as HTMLElement)?.tagName === 'LABEL'
```
Comment (holazz) [inline packages/components/checkbox/src/composables/use-checkbox-event.ts:59]: And here
```suggestion
        (item) => (item as HTMLElement).tagName === 'LABEL'
```

**Review by holazz — COMMENTED** [review]
Hi, @iDestin I left a few comments for your consideration. 😊

**Comment by iDestin** [comment]
> Hi, @iDestin I left a few comments for your consideration. 😊

ok! Has been modified

**Comment by holazz** [comment]
LGTM, Approved!

**Review by holazz — APPROVED** [review]

**Pull request merged**